--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -24,6 +24,7 @@ option(BUILD_SHARED "build warp-ctc shared library."  ON)
 option(WITH_ROCM     "Compile PaddlePaddle with ROCM platform"             OFF)

 if(WITH_ROCM)
+    list(APPEND CMAKE_PREFIX_PATH ${ROCM_PATH})
     add_definitions(-DWARPCTC_WITH_HIP)
     include(hip)
 endif(WITH_ROCM)

--- a/cmake/hip.cmake
+++ b/cmake/hip.cmake
@@ -64,8 +64,12 @@ set(HIP_CLANG_FLAGS ${HIP_CXX_FLAGS})
 # host linker to link.
 list(APPEND HIP_HCC_FLAGS -fno-gpu-rdc)
 list(APPEND HIP_HCC_FLAGS --amdgpu-target=gfx906)
+list(APPEND HIP_HCC_FLAGS --amdgpu-target=gfx926)
+list(APPEND HIP_HCC_FLAGS --amdgpu-target=gfx928)
 list(APPEND HIP_CLANG_FLAGS -fno-gpu-rdc)
 list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx906)
+list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx926)
+list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx928)
 
 
 if(HIP_COMPILER STREQUAL clang)